### ISSUE TITLE
vfs: intermittent test failure

### ISSUE THREAD

**Issue by ncw** [ncw/rclone#2083]
I saw this [in the travis logs](https://travis-ci.org/ncw/rclone/jobs/343530910#L949) in the non -race test.

Not sure whether it is related to your recent changes or not @B4dM4n ?

```
2018/02/19 20:14:53 NOTICE: Config file "/notfound" not found - using defaults
2018/02/19 20:14:53 ERROR : : error listing: directory not found
2018/02/19 20:14:53 ERROR : : error listing: directory not found
2018/02/19 20:14:53 ERROR : : error listing: directory not found
2018/02/19 20:14:53 ERROR : : error listing: directory not found
2018/02/19 20:14:53 ERROR : dir/: Can only open directories read only
2018/02/19 20:14:53 ERROR : dir/: Dir.Mkdir failed to create directory: mkdir /tmp/rclone784685176/dir/file1: not a directory
2018/02/19 20:14:53 ERROR : dir/: Dir.Remove not empty
2018/02/19 20:14:53 ERROR : dir/not found: Dir.Rename error: file does not exist
2018/02/19 20:14:53 ERROR : dir/file1: Can't figure out how to open with flags: 0x3
2018/02/19 20:14:53 ERROR : dir/file1: ReadFileHandle.Read error: Bad file descriptor
2018/02/19 20:14:53 ERROR : Error traversing cache "/home/travis/.cache/rclone/vfs/local/tmp/rclone784685176": lstat /home/travis/.cache/rclone/vfs/local/tmp/rclone784685176/file1: no such file or directory
2018/02/19 20:14:53 ERROR : Error traversing cache "/home/travis/.cache/rclone/vfs/local/tmp/rclone784685176": open /home/travis/.cache/rclone/vfs/local/tmp/rclone784685176: no such file or directory
2018/02/19 20:14:53 ERROR : open-test-file: File.openRW failed: open for read write: cache open file failed: open /home/travis/.cache/rclone/vfs/local/tmp/rclone784685176/open-test-file: no such file or directory
--- FAIL: TestRWFileHandleOpenTests (0.00s)
	run.go:169: Remote "Local file system at /tmp/rclone784685176", Local "Local file system at /tmp/rclone695535479", Modify Window "1ns"
	Error Trace:	read_write_test.go:474
			read_write_test.go:528
	Error:      	Received unexpected error:
	            	open /home/travis/.cache/rclone/vfs/local/tmp/rclone784685176/open-test-file: no such file or directory
	            	cache open file failed
	            	github.com/ncw/rclone/vfs.(*RWFileHandle).openPending
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write.go:156
	            	github.com/ncw/rclone/vfs.newRWFileHandle
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write.go:75
	            	github.com/ncw/rclone/vfs.(*File).openRW
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:328
	            	github.com/ncw/rclone/vfs.(*File).Open
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:448
	            	github.com/ncw/rclone/vfs.(*VFS).OpenFile
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/vfs.go:411
	            	github.com/ncw/rclone/vfs.testRWFileHandleOpenTest
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:473
	            	github.com/ncw/rclone/vfs.TestRWFileHandleOpenTests
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:528
	            	testing.tRunner
	            		/home/travis/.gimme/versions/go1.8.6.linux.amd64/src/testing/testing.go:657
	            	runtime.goexit
	            		/home/travis/.gimme/versions/go1.8.6.linux.amd64/src/runtime/asm_amd64.s:2197
	            	open for read write
	            	github.com/ncw/rclone/vfs.(*File).openRW
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:330
	            	github.com/ncw/rclone/vfs.(*File).Open
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:448
	            	github.com/ncw/rclone/vfs.(*VFS).OpenFile
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/vfs.go:411
	            	github.com/ncw/rclone/vfs.testRWFileHandleOpenTest
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:473
	            	github.com/ncw/rclone/vfs.TestRWFileHandleOpenTests
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:528
	            	testing.tRunner
	            		/home/travis/.gimme/versions/go1.8.6.linux.amd64/src/testing/testing.go:657
	            	runtime.goexit
	            		/home/travis/.gimme/versions/go1.8.6.linux.amd64/src/runtime/asm_amd64.s:2197
	Test:       	TestRWFileHandleOpenTests
	Messages:   	os.O_RDONLY
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle: Read: Can't read and write to file without --vfs-cache-mode >= minimal
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle: ReadAt: Can't read and write to file without --vfs-cache-mode >= minimal
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle: Truncate: Can't change size without --vfs-cache-mode >= writes
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle: Can't open for write without O_TRUNC on existing file without --vfs-cache-mode >= writes
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle.Write: can't seek in file without --vfs-cache-mode >= writes
2018/02/19 20:14:53 ERROR : file1: WriteFileHandle.Write: error: Bad file descriptor
FAIL
FAIL	github.com/ncw/rclone/vfs	0.085s
```

**Comment by B4dM4n** [comment]
Not sure if this error is caused by my changes. 

I found a logic error in truncate handling (which existed before my changes), but I'm not sure this is the cause, as the error message on my pc is different.
```go
-		if fh.flags&os.O_CREATE != 0 && fh.file.exists() {
+		if fh.flags&os.O_CREATE == 0 && fh.file.exists() {
			// create and empty file if it exists on the source
			cacheFileOpenFlags |= os.O_CREATE
		}
```
In the case of truncating, this check will create the empty cache file if a remote object exists.

PS: the link you provided does not lead to a failed job

Besides the error, do you think it is time to drop test support for go1.6? go1.10 was released a few days ago and the option to create subtests would be really nice. It is hard to figure out the real test case in the huge `openTests` table and I'm not sure which one failed. Using subtests we could name the cases or at least number them.

**Comment by B4dM4n** [comment]
Lately some AppVeyor builds are failing randomly during the cmount test:
- [1889](https://ci.appveyor.com/project/ncw/rclone/build/1889#L162)
- [1884](https://ci.appveyor.com/project/ncw/rclone/build/1884#L163)
- [1881](https://ci.appveyor.com/project/ncw/rclone/build/1881#L161)

I did not have the time too look into them, but in 2 of the 3 there is a file missing.

@ncw Did your reported error appear again on Travis?

I try to create a fix for it and for #2079 this weekend.

**Comment by ncw** [comment]
> I found a logic error in truncate handling (which existed before my changes)

I have fixed that in f57e92b9a5666052e887e87ce0e4bf78a87f1bdd also a really stupid bug on windows here c3d0f68923ce03402be668f1958cb67f97b8aad9

> PS: the link you provided does not lead to a failed job

Oh, I re-did the build which overwrite the log I think.

> Did your reported error appear again on Travis?

Here is a [possibly related one](https://travis-ci.org/ncw/rclone/jobs/346256567)

I spent some time writing more tests, and I can't replicate that bug - it is most annoying!

```
2018/02/26 12:31:40 NOTICE: Config file "/notfound" not found - using defaults
2018/02/26 12:31:40 ERROR : : error listing: directory not found
2018/02/26 12:31:40 ERROR : : error listing: directory not found
2018/02/26 12:31:40 ERROR : : error listing: directory not found
2018/02/26 12:31:40 ERROR : : error listing: directory not found
2018/02/26 12:31:40 ERROR : dir/: Can only open directories read only
2018/02/26 12:31:40 ERROR : dir/: Dir.Mkdir failed to create directory: mkdir /tmp/rclone188319816/dir/file1: not a directory
2018/02/26 12:31:40 ERROR : dir/: Dir.Remove not empty
2018/02/26 12:31:40 ERROR : dir/not found: Dir.Rename error: file does not exist
2018/02/26 12:31:40 ERROR : dir/file1: Can't figure out how to open with flags: 0x3
2018/02/26 12:31:40 ERROR : dir/file1: ReadFileHandle.Read error: Bad file descriptor
2018/02/26 12:31:40 ERROR : file1: File.openRW failed: open for read write: cache open file failed: open /home/travis/.cache/rclone/vfs/local/tmp/rclone188319816/file1: no such file or directory
--- FAIL: TestRWFileHandleReleaseWrite (0.00s)
	run.go:169: Remote "Local file system at /tmp/rclone188319816", Local "Local file system at /tmp/rclone831501575", Modify Window "1ns"
	Error Trace:	read_write_test.go:43
			read_write_test.go:441
	Error:      	Received unexpected error:
	            	open /home/travis/.cache/rclone/vfs/local/tmp/rclone188319816/file1: no such file or directory
	            	cache open file failed
	            	github.com/ncw/rclone/vfs.(*RWFileHandle).openPending
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write.go:156
	            	github.com/ncw/rclone/vfs.newRWFileHandle
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write.go:75
	            	github.com/ncw/rclone/vfs.(*File).openRW
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:328
	            	github.com/ncw/rclone/vfs.(*File).Open
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:448
	            	github.com/ncw/rclone/vfs.(*VFS).OpenFile
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/vfs.go:411
	            	github.com/ncw/rclone/vfs.rwHandleCreateWriteOnly
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:42
	            	github.com/ncw/rclone/vfs.TestRWFileHandleReleaseWrite
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:441
	            	testing.tRunner
	            		/home/travis/.gimme/versions/go1.10.linux.amd64/src/testing/testing.go:777
	            	runtime.goexit
	            		/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/asm_amd64.s:2361
	            	open for read write
	            	github.com/ncw/rclone/vfs.(*File).openRW
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:330
	            	github.com/ncw/rclone/vfs.(*File).Open
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/file.go:448
	            	github.com/ncw/rclone/vfs.(*VFS).OpenFile
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/vfs.go:411
	            	github.com/ncw/rclone/vfs.rwHandleCreateWriteOnly
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:42
	            	github.com/ncw/rclone/vfs.TestRWFileHandleReleaseWrite
	            		/home/travis/gopath/src/github.com/ncw/rclone/vfs/read_write_test.go:441
	            	testing.tRunner
	            		/home/travis/.gimme/versions/go1.10.linux.amd64/src/testing/testing.go:777
	            	runtime.goexit
	            		/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/asm_amd64.s:2361
	Test:       	TestRWFileHandleReleaseWrite
```

**Comment by ncw** [comment]
The only ones of these we are seeing regularly now are the TestDirCacheFlushOnDirRename - I've made a new issue about that #2202

SInce we aren't seeing any of the other failure modes any more, I'm going to close this.
